tpm/src/emulator.rs

```diff
@@ -0,0 +1,450 @@
+                cmd.output, cmd.output_len, isselftest
+            );
+
+            if isselftest {
+                let errcode: &[u8; 4] = cmd.output[6..6 + 4]
+                    .try_into()
+                    .expect("slice with incorrect length");
+                cmd.selftest_done = u32::from_ne_bytes(*errcode).to_be() == 0;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn deliver_request(&mut self, in_cmd: &mut BackendCmd) -> Result<Vec<u8>> {
+        if self.cmd.is_some() {
+            //previous request did not finish cleanly
+            return Err(Error::DeliverRequest(anyhow!(
+                "Cannot deliver tpm Request as previous cmd did not complete."
+            )));
+        }
+        self.cmd = Some(in_cmd.clone());
+
+        self.unix_tx_bufs()?;
+
+        let output = self.cmd.as_ref().unwrap().output.clone();
+        in_cmd.output.fill(0);
+        in_cmd.output.clone_from(&output);
+
+        self.backend_request_completed();
+        Ok(output)
+    }
+
+    pub fn backend_request_completed(&mut self) {
+        self.cmd = None;
+    }
+
+    pub fn cancel_cmd(&mut self) -> Result<()> {
+        let mut res: PtmResult = 0;
+
+        // Check if emulator implements Cancel Cmd
+        if (self.caps & PTM_CAP_CANCEL_TPM_CMD) != PTM_CAP_CANCEL_TPM_CMD {
+            return Err(Error::CheckCaps(anyhow!(
+                "Emulator does not implement 'Cancel Command' Capability"
+            )));
+        }
+        self.ctrlcmd(
+            Commands::CmdCancelTpmCmd,
+            &mut res,
+            0,
+            mem::size_of::<u32>(),
+        )?;
+        Ok(())
+    }
+
+    /// Configure buffersize to use while communicating with swtpm
+    fn set_buffer_size(&mut self, wantedsize: usize, actualsize: &mut usize) -> Result<()> {
+        let mut psbs: PtmSetBufferSize = PtmSetBufferSize::new(wantedsize as u32);
+
+        self.stop_tpm()?;
+
+        self.ctrlcmd(
+            Commands::CmdSetBufferSize,
+            &mut psbs,
+            mem::size_of::<u32>(),
+            4 * mem::size_of::<u32>(),
+        )?;
+
+        *actualsize = psbs.get_bufsize() as usize;
+
+        Ok(())
+    }
+
+    pub fn startup_tpm(&mut self, buffersize: usize) -> Result<()> {
+        let mut init: PtmInit = PtmInit::new();
+
+        let mut actual_size: usize = 0;
+
+        if buffersize != 0 {
+            self.set_buffer_size(buffersize, &mut actual_size)?;
+            debug!("set tpm buffersize to {:?} during Startup", buffersize);
+        }
+
+        self.ctrlcmd(
+            Commands::CmdInit,
+            &mut init,
+            mem::size_of::<u32>(),
+            mem::size_of::<u32>(),
+        )?;
+
+        Ok(())
+    }
+
+    fn stop_tpm(&mut self) -> Result<()> {
+        let mut res: PtmResult = 0;
+
+        self.ctrlcmd(Commands::CmdStop, &mut res, 0, mem::size_of::<u32>())?;
+
+        Ok(())
+    }
+
+    pub fn get_buffer_size(&mut self) -> Result<usize> {
```

Comment:
This is a strange way to do it. You should set a default buffer size in the constructor. Accessors shouldn't take "&mut self".

Reply:
This is defined by protocol. Below is the relevant excerpt from https://github.com/stefanberger/swtpm/blob/master/man/man3/swtpm_ioctls.pod
```
If a 0 is set in the buffer size of the request, the response will return the buffer size that is currently in use.
```